presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/UnaliasSymbolReferences.java

```diff
@@ -299,6 +302,61 @@ public PlanNode visitExchange(ExchangeNode node, RewriteContext<Void> context)
             return new ExchangeNode(node.getId(), node.getType(), node.getScope(), partitioningScheme, sources, inputs);
         }
 
+        private void mapExchangeNodeSymbols(ExchangeNode node)
+        {
+            if (node.getInputs().size() == 1) {
+                mapExchangeNodeOutputToInputSymbols(node);
+                return;
+            }
+
+            // Mapping from list [node.getInput(0).get(symbolIndex), node.getInput(1).get(symbolIndex), ...] to node.getOutputSymbols(symbolIndex).
+            // All symbols are canonical.
+            Map<List<Symbol>, Symbol> inputsToOutputs = new HashMap<>();
+            // Map each same list of input symbols [I1, I2, ..., In] to the same output symbol O
+            for (int symbolIndex = 0; symbolIndex < node.getOutputSymbols().size(); symbolIndex++) {
+                Symbol canonicalOutput = canonicalize(node.getOutputSymbols().get(symbolIndex));
+                List<Symbol> canonicalInputs = canonicalizeExchangeNodeInputs(node, symbolIndex);
+                Optional<Symbol> outputSymbolForInputSymbols = exchangeNodeOutputSymbolForInputSymbols(node, canonicalInputs, inputsToOutputs);
+
+                if (outputSymbolForInputSymbols.isPresent() && !canonicalOutput.equals(outputSymbolForInputSymbols.get())) {
+                    map(canonicalOutput, outputSymbolForInputSymbols.get());
+                }
+                else {
+                    inputsToOutputs.put(canonicalInputs, canonicalOutput);
+                }
+            }
+        }
+
+        private void mapExchangeNodeOutputToInputSymbols(ExchangeNode node)
+        {
+            checkState(node.getInputs().size() == 1);
+
+            for (int symbolIndex = 0; symbolIndex < node.getOutputSymbols().size(); symbolIndex++) {
+                Symbol canonicalOutput = canonicalize(node.getOutputSymbols().get(symbolIndex));
+                Symbol canonicalInput = canonicalize(node.getInputs().get(0).get(symbolIndex));
+
+                if (!canonicalOutput.equals(canonicalInput)) {
+                    map(canonicalOutput, canonicalInput);
+                }
+            }
+        }
+
+        private Optional<Symbol> exchangeNodeOutputSymbolForInputSymbols(ExchangeNode node, List<Symbol> canonicalInputs, Map<List<Symbol>, Symbol> inputsToOutputs)
+        {
+            if (node.getInputs().isEmpty() || !inputsToOutputs.containsKey(canonicalInputs)) {
```

Comment:
There is no assertion in `ExchangeNode` that is should be non-empty, so I decided to be on safe side.